### ISSUE TITLE
WiX4 Feature Request: Support external IHeatExtensions

### ISSUE THREAD

**Issue by InKahootz** [wixtoolset/issues#6564]
# Feature request

Major edit: I'm editing this to follow the feature template since the Twitch VODs show how you scorn people for not following the templates 🤣

## Describe the scenario and benefits that the feature supports

I’ve been building the WiX4-preview.1 locally so I can migrate my heat extension to WiXv4.
Everything has migrated relatively seamlessly and the targets hook fine into the `$(HarvestDependsOn)` property group.

Come to find out there's currently no possible way to get an `IHeatExtension` into the heat pipeline.

In the heat executable Run method, the `ICommandLineArguments` service parses my arguments and fills out the extensions property as expected. [wix4/src/wix/heat/Program.cs#L70](https://github.com/wixtoolset/wix4/blob/a4dc574fa13d828614058906c4e99fcb45965fe5/src/wix/heat/Program.cs#L70)

However, nothing is ever done with those extensions even when HeatCommandLineFactory.CreateCommandLine supports passing in an `IEnumerable<IHeatExtensions>` [wix4/src/wix/heat/Program.cs#L73](https://github.com/wixtoolset/wix4/blob/a4dc574fa13d828614058906c4e99fcb45965fe5/src/wix/heat/Program.cs#L73)

TL;DR; Support external HeatExtensions (`IHeatExtension` | `BaseHeatExtension` in WiX4) because they were supported in WiX3 even if they are niche.

## How I'm accomplishing this today

Pretty standard WiX3 extension. It has a custom target that runs right before the default harvest targets.  
My MSBuild task inherits from HeatTask (since I add some extra functionality to that. _very similar to HeatDirectory_). When the task runs `BuildCommandLine`, it passes the heat extension dll like so

```csharp

protected override void BuildCommandLine(WixCommandLineBuilder commandLineBuilder)
{
    // Append the new operation name
    commandLineBuilder.AppendSwitch(OperationName);

    commandLineBuilder.AppendFileNameIfNotNull(Directory);

    // include and exclude globs
    commandLineBuilder.AppendArrayIfNotNull("-include ", Include);
    commandLineBuilder.AppendArrayIfNotNull("-exclude ", Exclude);

    // The magic line that tells heat.exe to load our extension.
    commandLineBuilder.AppendSwitchIfNotNull("-ext ", ExtensionPath);

    // rest of switches here
    // <snip>

    base.BuildCommandLine(commandLineBuilder);
}

```

That extension assembly contains a class that inherits `HeatExtension` and also contains as assembly attribute of `[assembly: AssemblyDefaultHeatExtension(typeof(<HEAT_EXTENSION_CLASS>))]` so heat knows what to instantiate. You know the reset.

## What you'd like the new feature to do

Pretty much like how I’ve modified my fork so the `Run` method uses all the discovered extensions from the extension manager service.
[InKahootz/wix4/src/wix/heat/Program.cs#L77-L81](https://github.com/InKahootz/wix4/blob/d7b6e57b5d7d30fe71235267c0216ec2ff3e4ce8/src/wix/heat/Program.cs#L74-L81)
This works great! Just as I thought it would.  

I can't directly think of a way of using `@(WixExtension)` item group because you typically have to already write a custom MSBuild target to call your task.  
I think this would also call for moving most of the types in wix4/src/wix/heat/extensibility to WixToolset.Extensibility. It would also be great if you could make all the Harvester/Mutator types public again instead of internal (`FileHarvester`, `UtilHarvesterMutator`, etc...) since I used those in WiX3.

I can make my extension public if you would like to see how I’m doing things so far.

Thanks,  
Hank

P.S. Thanks for your work on WiX4

**Comment by rseanhall** [comment]
This really isn't the right place to discuss working on the toolset itself, it should be on the [wix-devs mailing list](https://wixtoolset.org/documentation/mailinglist/). You have to send an email to wix-devs-subscribe@lists.wixtoolset.org with the subject "Subscribe", or click "Preferences" at that web page and fill in the form there.

I'll leave this issue open since you're right, v4 Heat does not currently support external extensions. When implementing Heat for v4, I missed that `ICommandLineArguments` was preprocessing the extension arguments so I thought it was going to have to process a configuration file like in v3 in order to load them before parsing the command line. There were doubts on how many people were actually using Heat extensions so this functionality wasn't high on the priority list.

**Comment by InKahootz** [comment]
Appreciate the reply.

I can take further discussion to the devs list. I am already subscribed to that as well.

**Comment by ericstj** [comment]
Just a bystander here, but I noticed the Heat task supports [AdditionalOptions](https://github.com/wixtoolset/wix4/blob/a4dc574fa13d828614058906c4e99fcb45965fe5/src/wix/WixToolset.BuildTasks/ToolsetTask.cs#L21) as a pass-through so that you can plumb any additional command line arguments.  That could help you avoid deriving from HeatTask, since you can just reuse the one WiX provides and write your target that builds up a string to pass to AdditionalOptions based on the things you wish to add.

**Comment by InKahootz** [comment]
@ericstj  
HeatTask is abstract so any of the deriving classes you try to use will put their own directive first (e.g `dir`, `file`, or `project`) so that ruins those classes. You could override the `BuildCommandLine` but that requires having a reference to some wix dlls in the buildtask project which I explain why I can't do that later.

Some other things I ran into while rewriting my extension.

Previously, my extension was one project but I had to split it into multiple. One that has build tasks and the other that has the extension implementation, and then one other project that just pulls those two together in a nuget package.  
The BuildTasks project isn't allowed to reference any WiX in my case (specifically I wanted to use `ToolsetTask` for finding the heat tool) but when MSBuild runs in .NETCore runtime, the dlls are completely isolated, (`AssemblyLoadContext`) to avoid dependency tree collision. aka. DLL hell. All the wix dlls have to sit right beside the build task dll which ruins some of the ToolsetTask reflection and setting the ToolTask.Toolpath caused other problems.

In .NETFramework MSBuild, you can expect the WixToolset.BuildTasks to kind of side-loaded from the wix dir and reflection on where the tools were worked nicely. But I need it to work for both runtimes at the same time.

The actual extension project can reference WiX dlls because the extensions are not loaded in isolation in either of the frameworks.

There's many other little things I learned but it's hard to write it all down in one comment.
Feel free to browse my extension here: https://github.com/InKahootz/HarvestDirectoryEx
